crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -214,76 +213,96 @@ where
+            last_segment_header.segment_index() - SegmentIndex::ONE,
             self.piece_getter,
             &mut reconstructor,
         )
         .await?;
 
-        let blocks_in_second_last_segment = second_last_segment_blocks.len();
-        debug!(
-            "Second last segment blocks downloaded (SegmentId={}): {}-{}",
-            second_last_segment_index,
-            second_last_segment_blocks[0].0,
-            second_last_segment_blocks[blocks_in_second_last_segment - 1].0
-        );
-
         let blocks = download_and_reconstruct_blocks(
             last_segment_header.segment_index(),
             self.piece_getter,
             &mut reconstructor,
         )
         .await?;
+        let mut blocks = VecDeque::from(blocks);
 
-        let blocks_in_last_segment = blocks.len();
-        debug!(
-            "Blocks downloaded (SegmentId={}): {}-{}",
-            last_segment_index,
-            blocks[0].0,
-            blocks[blocks_in_last_segment - 1].0
-        );
-
-        // 2. Raw import the last block from the second last segment.
+        let (first_block_number, first_block_bytes) = blocks
+            .pop_front()
+            // TODO: Ensure this expectation actually holds, currently this is not guaranteed
+            .expect("List of blocks is not empty according to logic above; qed");
 
-        // 3. Download state for the first block of the last segment.
+        let last_block_number = blocks
+            .back()
+            .map_or(first_block_number, |(block_number, _block_bytes)| {
+                *block_number
+            });
 
-        let (second_last_block_number, _) = blocks[blocks_in_last_segment - 2].clone();
-        let last_block = blocks[blocks_in_last_segment - 1].clone();
-        let last_block_number = last_block.0;
+        debug!(
+            %last_segment_index,
+            %first_block_number,
+            %last_block_number,
+            "Blocks from last segment downloaded"
+        );
 
-        // Raw import state block and download state.
-        // Add the first block of the last segment as raw
+        let signed_block =
+            decode_block::<Block>(&first_block_bytes).map_err(|error| error.to_string())?;
 
-        // The state block is first block from the last segment.
-        let state_block_bytes = blocks[0].1.clone();
-        let state_block_number = blocks[0].0;
+        let (header, extrinsics) = signed_block.block.deconstruct();
+        let block_hash = header.hash();
 
-        let (raw_block, _) =
-            Self::create_raw_block(state_block_bytes.clone(), state_block_number.into())?;
+        debug!(
+            %first_block_number,
+            ?block_hash,
+            parent_hash = ?header.parent_hash(),
+            "Reconstructed block for raw block import"
+        );
 
-        import_raw_block(self.client.as_ref(), raw_block.clone())?;
+        let mut import_block = BlockImportParams::new(
+            if blocks.is_empty() {
```

Comment:
Why did you introduce branches here?

Reply:
If `blocks` is empty then naturally the block we insert directly bypassing block import is the first and only block we are going to import, hence we should use `BlockOrigin::NetworkBroadcast` to notify Substrate's sync mechanism about it as we do below.